diff --git a/src/api/CMakeLists.txt b/src/api/CMakeLists.txt
index 4a5514a7b..1c5a2c033 100644
--- a/src/api/CMakeLists.txt
+++ b/src/api/CMakeLists.txt
@@ -63,6 +63,7 @@ z3_add_component(api
     api_solver.cpp
     api_stats.cpp
     api_tactic.cpp
+    api_special_relations.cpp # ASHU added
     z3_replayer.cpp
     ${full_path_generated_files}
   COMPONENT_DEPENDENCIES
diff --git a/src/api/api_ast.cpp b/src/api/api_ast.cpp
index 536b94fb2..f0887c1fa 100644
--- a/src/api/api_ast.cpp
+++ b/src/api/api_ast.cpp
@@ -1203,6 +1203,19 @@ extern "C" {
             }
         }
 
+        //BEGIN: ASHU
+        if (mk_c(c)->get_special_relations_fid() == _d->get_family_id()) {
+            switch(_d->get_decl_kind()) {
+            case OP_SPECIAL_RELATION_LO : return Z3_OP_SPECIAL_RELATION_LO;
+            case OP_SPECIAL_RELATION_PO : return Z3_OP_SPECIAL_RELATION_PO;
+            case OP_SPECIAL_RELATION_PO_AO : return Z3_OP_SPECIAL_RELATION_PO_AO;
+            case OP_SPECIAL_RELATION_PLO: return Z3_OP_SPECIAL_RELATION_PLO;
+            case OP_SPECIAL_RELATION_TO : return Z3_OP_SPECIAL_RELATION_TO;
+            default: UNREACHABLE();
+            }
+        }
+        //END: ASHU
+
         return Z3_OP_UNINTERPRETED;
         Z3_CATCH_RETURN(Z3_OP_UNINTERPRETED);
     }
diff --git a/src/api/api_context.cpp b/src/api/api_context.cpp
index 5993e9fdd..5a5ed9344 100644
--- a/src/api/api_context.cpp
+++ b/src/api/api_context.cpp
@@ -79,6 +79,7 @@ namespace api {
         m_datalog_util(m()),
         m_fpa_util(m()),
         m_sutil(m()),
+        m_sr_util(m()),  //ASHU: addition
         m_last_result(m()),
         m_ast_trail(m()),
         m_pmanager(m_limit) {
@@ -100,6 +101,7 @@ namespace api {
         m_datalog_fid = m().mk_family_id("datalog_relation");
         m_fpa_fid   = m().mk_family_id("fpa");
         m_seq_fid   = m().mk_family_id("seq");
+        m_sr_fid    = m().mk_family_id("special_relations"); //ASHU: addition
         m_dt_plugin = static_cast<datatype_decl_plugin*>(m().get_plugin(m_dt_fid));
     
         install_tactics(*this);
diff --git a/src/api/api_context.h b/src/api/api_context.h
index 50e89113d..57fd8edd9 100644
--- a/src/api/api_context.h
+++ b/src/api/api_context.h
@@ -29,6 +29,7 @@ Revision History:
 #include "ast/datatype_decl_plugin.h"
 #include "ast/dl_decl_plugin.h"
 #include "ast/fpa_decl_plugin.h"
+#include "ast/special_relations_decl_plugin.h" //ASHU: addition
 #include "smt/smt_kernel.h"
 #include "smt/params/smt_params.h"
 #include "util/event_handler.h"
@@ -62,6 +63,7 @@ namespace api {
         datalog::dl_decl_util      m_datalog_util;
         fpa_util                   m_fpa_util;
         seq_util                   m_sutil;
+        special_relations_util     m_sr_util; //ASHU: addition
 
         // Support for old solver API
         smt_params                 m_fparams;
@@ -83,6 +85,7 @@ namespace api {
         family_id                  m_pb_fid;
         family_id                  m_fpa_fid;
         family_id                  m_seq_fid;
+        family_id                  m_sr_fid; //ASHU: addition
         datatype_decl_plugin *     m_dt_plugin;
         
         std::string                m_string_buffer; // temporary buffer used to cache strings sent to the "external" world.
@@ -128,6 +131,7 @@ namespace api {
         fpa_util & fpautil() { return m_fpa_util; }
         datatype_util& dtutil() { return m_dt_plugin->u(); }
         seq_util& sutil() { return m_sutil; }
+        special_relations_util& sr_util() { return m_sr_util; } //ASHU: added
         family_id get_basic_fid() const { return m_basic_fid; }
         family_id get_array_fid() const { return m_array_fid; }
         family_id get_arith_fid() const { return m_arith_fid; }
@@ -137,6 +141,7 @@ namespace api {
         family_id get_pb_fid() const { return m_pb_fid; }
         family_id get_fpa_fid() const { return m_fpa_fid; }
         family_id get_seq_fid() const { return m_seq_fid; }
+        family_id get_special_relations_fid() const { return m_sr_fid; } //ASHU: added
         datatype_decl_plugin * get_dt_plugin() const { return m_dt_plugin; }
 
         Z3_error_code get_error_code() const { return m_error_code; }
diff --git a/src/api/api_solver.cpp b/src/api/api_solver.cpp
index 657ff025c..2d8569b40 100644
--- a/src/api/api_solver.cpp
+++ b/src/api/api_solver.cpp
@@ -45,6 +45,12 @@ Revision History:
 #include "sat/sat_solver.h"
 #include "sat/tactic/goal2sat.h"
 
+//BEGIN: ASHU
+Z3_ast_vector_ref* asserted_linear_constr = NULL;
+ast_ref_vector* asserted_linear_constr_vector;
+Z3_ast_vector_ref* asserted_po_constr = NULL;
+ast_ref_vector* asserted_po_constr_vector;
+//END: ASHU
 
 extern "C" {
 
@@ -60,6 +66,18 @@ extern "C" {
         context_params::collect_solver_param_descrs(r);
         p.validate(r);
         s->m_solver->updt_params(p);
+        //BEGIN: ASHU
+        if (asserted_linear_constr == NULL) {
+          asserted_linear_constr = alloc(Z3_ast_vector_ref, *mk_c(c), mk_c(c)->m()); //alloc(Z3_ast_vector_ref, mk_c(c)->m());
+          asserted_linear_constr_vector = &(asserted_linear_constr->m_ast_vector);
+          asserted_linear_constr->inc_ref();
+        }
+        if (asserted_po_constr == NULL) {
+          asserted_po_constr = alloc(Z3_ast_vector_ref, *mk_c(c), mk_c(c)->m());
+          asserted_po_constr_vector = &(asserted_po_constr->m_ast_vector);
+          asserted_po_constr->inc_ref();
+        }
+        //END: ASHU
     }
 
     static void init_solver(Z3_context c, Z3_solver s) {
@@ -628,5 +646,24 @@ extern "C" {
         Z3_CATCH_RETURN(0);        
     }
 
+  //BEGIN: ASHU
+
+  Z3_ast_vector Z3_API Z3_collect_last_asserted_linear_constr(void) {
+    return of_ast_vector(asserted_linear_constr);
+  }
+
+  void Z3_API Z3_delete_last_asserted_linear_constr(void) {
+    if (asserted_linear_constr != NULL)
+      asserted_linear_constr->dec_ref();
+  }
+  Z3_ast_vector Z3_API Z3_collect_last_asserted_po_constr(void) {
+    return of_ast_vector(asserted_po_constr);
+  }
+
+  void Z3_API Z3_delete_last_asserted_po_constr(void) {
+    if (asserted_po_constr != NULL)
+      asserted_po_constr->dec_ref();
+  }
+  //END: ASHU
 
 };
diff --git a/src/api/c++/z3++.h b/src/api/c++/z3++.h
index 665ffb438..27d766baa 100644
--- a/src/api/c++/z3++.h
+++ b/src/api/c++/z3++.h
@@ -153,7 +153,7 @@ namespace z3 {
     public:
         context() { config c; init(c); }
         context(config & c) { init(c); }
-        ~context() { Z3_del_context(m_ctx); }
+      ~context() { Z3_delete_last_asserted_linear_constr(); Z3_delete_last_asserted_po_constr(); Z3_del_context(m_ctx); } // ASHU
         operator Z3_context() const { return m_ctx; }
 
         /**
@@ -286,6 +286,9 @@ namespace z3 {
 
         expr constant(symbol const & name, sort const & s);
         expr constant(char const * name, sort const & s);
+        //BEGIN: ASHU
+        expr fresh_constant(char const * name, sort const & s);// ASHU
+       //END: ASHU
         expr bool_const(char const * name);
         expr int_const(char const * name);
         expr real_const(char const * name);
@@ -318,6 +321,11 @@ namespace z3 {
 
         expr num_val(int n, sort const & s);
 
+      //BEGIN: ASHU
+      expr_vector collect_last_asserted_linear_constr(void);
+      expr_vector collect_last_asserted_po_constr(void);
+      //END: ASHU
+
         /**
            \brief parsing
          */
@@ -595,6 +603,13 @@ namespace z3 {
         expr(expr const & n):ast(n) {}
         expr & operator=(expr const & n) { return static_cast<expr&>(ast::operator=(n)); }
 
+      //BEGIN: ASHU
+        bool get_bool() const {
+          Z3_lbool b = Z3_get_bool_value(*m_ctx, m_ast);
+          assert(b!=Z3_L_UNDEF);
+          return b==Z3_L_TRUE;
+        }
+      //END: ASHU
         /**
            \brief Return the sort of this expression.
         */
@@ -2215,6 +2230,14 @@ namespace z3 {
         }
     public:
         tactic(context & c, char const * name):object(c) { Z3_tactic r = Z3_mk_tactic(c, name); check_error(); init(r); }
+        tactic(tactic const & tactic, params const & params):object(tactic) {
+          Z3_param_descrs pd = Z3_tactic_get_param_descrs(ctx(), tactic);
+          Z3_param_descrs_inc_ref(ctx(), pd);
+          Z3_params_validate(ctx(), params, pd);
+          check_error();
+          Z3_param_descrs_dec_ref(ctx(), pd);
+          Z3_tactic r = Z3_tactic_using_params(tactic.ctx(), tactic, params); check_error(); init(r);
+        }
         tactic(context & c, Z3_tactic s):object(c) { init(s); }
         tactic(tactic const & s):object(s) { init(s.m_tactic); }
         ~tactic() { Z3_tactic_dec_ref(ctx(), m_tactic); }
@@ -2630,6 +2653,13 @@ namespace z3 {
         check_error();
         return expr(*this, r);
     }
+  //BEGIN: ASHU
+    inline expr context::fresh_constant(char const * name, sort const & s) {
+      Z3_ast r = Z3_mk_fresh_const(m_ctx, name, s);
+      check_error();
+      return expr(*this, r);
+    }
+  //END: ASHU
     inline expr context::constant(char const * name, sort const & s) { return constant(str_symbol(name), s); }
     inline expr context::bool_const(char const * name) { return constant(name, bool_sort()); }
     inline expr context::int_const(char const * name) { return constant(name, int_sort()); }
@@ -2955,6 +2985,40 @@ namespace z3 {
 
 
 
+    inline expr sr_lo(expr const& a, expr const& b) {
+        check_context(a, b);
+        Z3_ast r = Z3_mk_sr_lo(a.ctx(), a, b);
+        a.check_error();
+        return expr(a.ctx(), r);
+    }
+
+    inline expr sr_po(expr const& a, expr const& b) {
+        check_context(a, b);
+        Z3_ast r = Z3_mk_sr_po(a.ctx(), a, b);
+        a.check_error();
+        return expr(a.ctx(), r);
+    }
+
+    inline expr sr_po_ao(expr const& a, expr const& b) {
+        check_context(a, b);
+        Z3_ast r = Z3_mk_sr_po_ao(a.ctx(), a, b);
+        a.check_error();
+        return expr(a.ctx(), r);
+    }
+
+    inline expr sr_plo(expr const& a, expr const& b) {
+        check_context(a, b);
+        Z3_ast r = Z3_mk_sr_plo(a.ctx(), a, b);
+        a.check_error();
+        return expr(a.ctx(), r);
+    }
+
+    inline expr sr_to(expr const& a, expr const& b) {
+        check_context(a, b);
+        Z3_ast r = Z3_mk_sr_to(a.ctx(), a, b);
+        a.check_error();
+        return expr(a.ctx(), r);
+    }
 
     inline expr_vector context::parse_string(char const* s) {
         Z3_ast_vector r = Z3_parse_smtlib2_string(*this, s, 0, 0, 0, 0, 0, 0);
@@ -3025,7 +3089,29 @@ namespace z3 {
         return expr(ctx(), r);
     }
 
-
+  //BEGIN: ASHU
+    inline expr_vector context::collect_last_asserted_linear_constr(void) {
+      Z3_ast_vector asserted1 = Z3_collect_last_asserted_linear_constr();
+      z3::expr_vector asserted = expr_vector(*this, asserted1);
+      return asserted;
+    }
+
+    inline expr_vector context::collect_last_asserted_po_constr(void) {
+      Z3_ast_vector asserted1 = Z3_collect_last_asserted_po_constr();
+      z3::expr_vector asserted = expr_vector(*this, asserted1);
+      return asserted;
+    }
+    // inline expr distinct(expr_vector const& args) {
+    //   array<Z3_ast> _args(args.size());
+    //   for (unsigned i = 0; i < args.size(); i++) {
+    //     check_context(args.ctx(), args[i]);
+    //     _args[i] = args[i];
+    //   }
+    //   Z3_ast r = Z3_mk_distinct(args.ctx(), args.size(), _args.ptr());
+    //   args.ctx().check_error();
+    //   return expr(args.ctx(), r);
+    // }
+  //END: ASHU
 
 }
 
diff --git a/src/api/z3_api.h b/src/api/z3_api.h
index 3e47833b4..6609ebe85 100644
--- a/src/api/z3_api.h
+++ b/src/api/z3_api.h
@@ -1257,6 +1257,14 @@ typedef enum {
 
     Z3_OP_INTERNAL,
 
+    //BEGIN: ASHU
+    Z3_OP_SPECIAL_RELATION_LO,
+    Z3_OP_SPECIAL_RELATION_PO,
+    Z3_OP_SPECIAL_RELATION_PO_AO,
+    Z3_OP_SPECIAL_RELATION_PLO,
+    Z3_OP_SPECIAL_RELATION_TO,
+    //END: ASHU
+
     Z3_OP_UNINTERPRETED
 } Z3_decl_kind;
 
@@ -3505,6 +3513,52 @@ extern "C" {
     /*@}*/
 
 
+    /**
+       \brief declare \c a and \c b are in linear order.
+
+       \pre a and b are of same type.
+
+       def_API('Z3_mk_sr_lo' ,AST ,(_in(CONTEXT), _in(AST), _in(AST)))
+     */
+    Z3_ast Z3_API Z3_mk_sr_lo(Z3_context c, Z3_ast a, Z3_ast b);
+
+    /**
+       \brief declare \c a and \c b are in partial order.
+
+       \pre a and b are of same type.
+
+       def_API('Z3_mk_sr_po' ,AST ,(_in(CONTEXT), _in(AST), _in(AST)))
+     */
+    Z3_ast Z3_API Z3_mk_sr_po(Z3_context c, Z3_ast a, Z3_ast b);
+
+    /**
+       \brief declare \c a and \c b are already partial ordered.
+
+       \pre a and b are of same type.
+
+       def_API('Z3_mk_sr_po_ao' ,AST ,(_in(CONTEXT), _in(AST), _in(AST)))
+     */
+    Z3_ast Z3_API Z3_mk_sr_po_ao(Z3_context c, Z3_ast a, Z3_ast b);
+
+    /**
+       \brief declare \c a and \c b are in piecewise linear order.
+
+       \pre a and b are of same type.
+
+       def_API('Z3_mk_sr_plo' ,AST ,(_in(CONTEXT), _in(AST), _in(AST)))
+     */
+    Z3_ast Z3_API Z3_mk_sr_plo(Z3_context c, Z3_ast a, Z3_ast b);
+
+    /**
+       \brief declare \c a and \c b are in total order.
+
+       \pre a and b are of same type.
+
+       def_API('Z3_mk_sr_to' ,AST ,(_in(CONTEXT), _in(AST), _in(AST)))
+     */
+    Z3_ast Z3_API Z3_mk_sr_to(Z3_context c, Z3_ast a, Z3_ast b);
+
+
     /** @name Quantifiers */
     /*@{*/
     /**
@@ -6285,6 +6339,13 @@ extern "C" {
 
     /*@}*/
 
+  //BEGIN: ASHU
+  Z3_ast_vector Z3_API Z3_collect_last_asserted_linear_constr(void);
+  void Z3_API Z3_delete_last_asserted_linear_constr(void);
+  Z3_ast_vector Z3_API Z3_collect_last_asserted_po_constr(void);
+  void Z3_API Z3_delete_last_asserted_po_constr(void);
+  //END: ASHU
+
 #ifdef __cplusplus
 }
 #endif // __cplusplus
diff --git a/src/ast/CMakeLists.txt b/src/ast/CMakeLists.txt
index 4dcdd2a35..3c06383f5 100644
--- a/src/ast/CMakeLists.txt
+++ b/src/ast/CMakeLists.txt
@@ -38,6 +38,7 @@ z3_add_component(ast
     pp.cpp
     reg_decl_plugins.cpp
     seq_decl_plugin.cpp
+    special_relations_decl_plugin.cpp # ASHU added
     shared_occs.cpp
     static_features.cpp
     used_vars.cpp
diff --git a/src/ast/reg_decl_plugins.cpp b/src/ast/reg_decl_plugins.cpp
index 985ecee9e..82a9fb372 100644
--- a/src/ast/reg_decl_plugins.cpp
+++ b/src/ast/reg_decl_plugins.cpp
@@ -26,6 +26,7 @@ Revision History:
 #include "ast/seq_decl_plugin.h"
 #include "ast/pb_decl_plugin.h"
 #include "ast/fpa_decl_plugin.h"
+#include "ast/special_relations_decl_plugin.h" //ASHU: added
 
 void reg_decl_plugins(ast_manager & m) {
     if (!m.get_plugin(m.mk_family_id(symbol("arith")))) {
@@ -52,4 +53,7 @@ void reg_decl_plugins(ast_manager & m) {
     if (!m.get_plugin(m.mk_family_id(symbol("pb")))) {
         m.register_plugin(symbol("pb"), alloc(pb_decl_plugin));
     }
+    if (!m.get_plugin(m.mk_family_id(symbol("special_relations")))) { //ASHU: added
+        m.register_plugin(symbol("special_relations"), alloc(special_relations_decl_plugin));
+    }
 }
diff --git a/src/cmd_context/cmd_context.cpp b/src/cmd_context/cmd_context.cpp
index 0bb4b091e..fe4ae9697 100644
--- a/src/cmd_context/cmd_context.cpp
+++ b/src/cmd_context/cmd_context.cpp
@@ -30,6 +30,7 @@ Notes:
 #include "ast/seq_decl_plugin.h"
 #include "ast/pb_decl_plugin.h"
 #include "ast/fpa_decl_plugin.h"
+#include "ast/special_relations_decl_plugin.h"
 #include "ast/ast_pp.h"
 #include "ast/rewriter/var_subst.h"
 #include "ast/pp.h"
@@ -694,6 +695,7 @@ void cmd_context::init_manager_core(bool new_manager) {
         register_plugin(symbol("pb"),       alloc(pb_decl_plugin), logic_has_pb());
         register_plugin(symbol("fpa"),      alloc(fpa_decl_plugin), logic_has_fpa());
         register_plugin(symbol("datalog_relation"), alloc(datalog::dl_decl_plugin), !has_logic());
+        register_plugin(symbol("special_relations"), alloc(special_relations_decl_plugin), !has_logic());
     }
     else {
         // the manager was created by an external module
diff --git a/src/smt/CMakeLists.txt b/src/smt/CMakeLists.txt
index fb1997fb4..ed2f185be 100644
--- a/src/smt/CMakeLists.txt
+++ b/src/smt/CMakeLists.txt
@@ -34,6 +34,7 @@ z3_add_component(smt
     smt_model_checker.cpp
     smt_model_finder.cpp
     smt_model_generator.cpp
+    smt_model_reporter.cpp # ASHU added
     smt_quantifier.cpp
     smt_quantifier_stat.cpp
     smt_quick_checker.cpp
@@ -62,6 +63,7 @@ z3_add_component(smt
     theory_str.cpp
     theory_utvpi.cpp
     theory_wmaxsat.cpp
+    theory_special_relations.cpp # ASHU added
     uses_theory.cpp
     watch_list.cpp
   COMPONENT_DEPENDENCIES
diff --git a/src/smt/diff_logic.h b/src/smt/diff_logic.h
index 44e858219..1f28c5907 100644
--- a/src/smt/diff_logic.h
+++ b/src/smt/diff_logic.h
@@ -296,6 +296,9 @@ public:
 
     numeral const& get_weight(edge_id id) const { return m_edges[id].get_weight(); }
 
+    edge_id_vector const& get_out_edges(dl_var v) const { return m_out_edges[v]; }
+
+    edge_id_vector const& get_in_edges(dl_var v) const { return m_in_edges[v]; }
 
 private:
     // An assignment is almost feasible if all but edge with idt edge are feasible.
@@ -739,7 +742,7 @@ public:
         if (idx2 < idx1) {
             std::swap(idx1,idx2);
         }        
-        (void) max_idx;
+        // (void) max_idx;
         SASSERT(idx1 < idx2 && idx2 < edges.size());
         SASSERT(max_idx < edges.size());
         dst = get_source(edges[idx2]);
@@ -1016,7 +1019,39 @@ public:
         threads[prev] = start;
     }
 
-    void bfs_undirected(dl_var start, svector<dl_var> & parents, svector<dl_var> & depths) {
+    bool can_reach(dl_var src, dl_var dst) {
+        uint_set target, visited;
+        target.insert(dst);
+        return reachable(src, target, visited, dst);
+    }
+
+    bool reachable(dl_var start, uint_set const& target, uint_set& visited, dl_var& dst) {
+        visited.reset();
+        svector<dl_var> nodes;
+        nodes.push_back(start);
+        for (unsigned i = 0; i < nodes.size(); ++i) {
+            dl_var n = nodes[i];
+            if (visited.contains(n)) continue;
+            visited.insert(n);
+            edge_id_vector & edges = m_out_edges[n];
+            typename edge_id_vector::iterator it  = edges.begin();
+            typename edge_id_vector::iterator end = edges.end();
+            for (; it != end; ++it) {
+                edge_id e_id = *it;
+                edge & e     = m_edges[e_id];
+                if (e.is_enabled()) {
+                    dst = e.get_target();
+                    if (target.contains(dst)) {
+                        return true;
+                    }
+                    nodes.push_back(dst);
+                }
+            }
+        }
+        return false;
+    }
+
+    void bfs_undirected(dl_var start, svector<dl_var> & parents, svector<unsigned> & depths) {
         parents.reset();
         parents.resize(get_num_nodes());
         parents[start] = -1;
@@ -1276,6 +1311,15 @@ public:
     // Return true if the path exists, false otherwise.
     template<typename Functor>
     bool find_shortest_zero_edge_path(dl_var source, dl_var target, unsigned timestamp, Functor & f) {
+      return find_shortest_path_aux(source, target, timestamp, f, true);
+    }
+    template<typename Functor>
+    bool find_shortest_reachable_path(dl_var source, dl_var target, unsigned timestamp, Functor & f) {
+      return find_shortest_path_aux(source, target, timestamp, f, false);
+    }
+
+    template<typename Functor>
+    bool find_shortest_path_aux(dl_var source, dl_var target, unsigned timestamp, Functor & f, bool zero_edge) {
         svector<bfs_elem> bfs_todo;
         svector<char>     bfs_mark;
         bfs_mark.resize(m_assignment.size(), false);
@@ -1303,7 +1347,7 @@ public:
                 }
                 set_gamma(e, gamma);
                 TRACE("dl_bfs", tout << "processing edge: "; display_edge(tout, e); tout << "gamma: " << gamma << "\n";);
-                if (gamma.is_zero() && e.get_timestamp() < timestamp) {
+                if ((gamma.is_zero() || (!zero_edge && gamma.is_neg())) && e.get_timestamp() < timestamp) {
                     dl_var curr_target = e.get_target();
                     TRACE("dl_bfs", tout << "curr_target: " << curr_target << 
                           ", mark: " << static_cast<int>(bfs_mark[curr_target]) << "\n";);
@@ -1352,6 +1396,27 @@ public:
         return false;
     } 
 
+  //begin: ASHU
+
+// Each node
+//  - 0 == "unreachable",
+//  - 1 == "forward blocked reachable"
+//  - 2 == "both ways blcoked reachable"
+//  - 3 == "reachable"
+
+enum reach_mark {
+  UNREACH = 0,
+  FWD_BLOCKED_REACH = 1,
+  REV_BLOCKED_REACH = 2,
+  REACH = 3
+};
+
+  template<typename Functor>
+  bool no_path( dl_var source, dl_var target, unsigned timestamp, Functor & f );
+    template<typename Functor>
+  bool no_path_both_way( dl_var source, dl_var target, unsigned timestamp, Functor & f );
+  void dotty_dump(dl_var source, dl_var target, edge_id_vector& boundry);
+//end: ASHU
 
     //
     // Theory propagation:
diff --git a/src/smt/smt_context.cpp b/src/smt/smt_context.cpp
index 14a19f824..fac67cae7 100644
--- a/src/smt/smt_context.cpp
+++ b/src/smt/smt_context.cpp
@@ -4371,6 +4371,15 @@ namespace smt {
         }
     }
 
+  //BEGIN: ASHU
+  void context::collect_asserted_linear_constr( ast_ref_vector& atoms ) {
+    m_model_generator->collect_asserted_linear_constr( atoms );
+  }
+  void context::collect_asserted_po_constr( ast_ref_vector& atoms ) {
+    m_model_generator->collect_asserted_po_constr( atoms );
+  }
+  //END: ASHU
+
 };
 
 
diff --git a/src/smt/smt_context.h b/src/smt/smt_context.h
index 32379c353..132d4ce2e 100644
--- a/src/smt/smt_context.h
+++ b/src/smt/smt_context.h
@@ -1585,6 +1585,11 @@ namespace smt {
         func_decl * get_macro_interpretation(unsigned i, expr_ref & interp) const { return m_asserted_formulas.get_macro_interpretation(i, interp); }
         quantifier * get_macro_quantifier(func_decl * f) const { return m_asserted_formulas.get_macro_quantifier(f); }
         void insert_macro(func_decl * f, quantifier * m, proof * pr, expr_dependency * dep) { m_asserted_formulas.insert_macro(f, m, pr, dep); }
+
+      //BEGIN: ASHU
+      void collect_asserted_linear_constr( ast_ref_vector& );
+      void collect_asserted_po_constr( ast_ref_vector& );
+      //END: ASHU
     };
 
 };
diff --git a/src/smt/smt_kernel.cpp b/src/smt/smt_kernel.cpp
index a6413aef9..cde35fb21 100644
--- a/src/smt/smt_kernel.cpp
+++ b/src/smt/smt_kernel.cpp
@@ -387,4 +387,13 @@ namespace smt {
         return m_imp->m_kernel;
     }
 
+  //BEGIN: ASHU
+  void kernel::collect_asserted_linear_constr( ast_ref_vector& atoms ) {
+    m_imp->m_kernel.collect_asserted_linear_constr( atoms );
+  }
+  void kernel::collect_asserted_po_constr( ast_ref_vector& atoms ) {
+    m_imp->m_kernel.collect_asserted_po_constr( atoms );
+  }
+  //END: ASHU
+
 };
diff --git a/src/smt/smt_kernel.h b/src/smt/smt_kernel.h
index 7b2f774ad..cbb6bc438 100644
--- a/src/smt/smt_kernel.h
+++ b/src/smt/smt_kernel.h
@@ -268,6 +268,10 @@ namespace smt {
            \warning We should not use this method
         */
         context & get_context();
+      //BEGIN: ASHU
+      void collect_asserted_linear_constr( ast_ref_vector& );
+      void collect_asserted_po_constr( ast_ref_vector& );
+      //END: ASHU
     };
 };
 
diff --git a/src/smt/smt_model_generator.cpp b/src/smt/smt_model_generator.cpp
index 9ba2c6165..70a480aec 100644
--- a/src/smt/smt_model_generator.cpp
+++ b/src/smt/smt_model_generator.cpp
@@ -32,7 +32,14 @@ namespace smt {
         m_context(nullptr),
         m_fresh_idx(1),
         m_asts(m_manager),
-        m_model(nullptr) {
+        m_model(nullptr),
+        //BEGIN: ASHU
+        f1s(m_manager),
+        f2s(m_manager),
+        added_constr(m_manager),
+        assigned_atoms(m_manager)
+        //END: ASHU
+    {
     }
 
     model_generator::~model_generator() {
@@ -311,20 +318,24 @@ namespace smt {
               svector<source>::const_iterator it  = sources.begin();
               svector<source>::const_iterator end = sources.end();
               for (; it != end; ++it) {
-                  source const & curr = *it;
-                  if (curr.is_fresh_value()) {
-                      tout << "fresh!" << curr.get_value()->get_idx() << " " << mk_pp(curr.get_value()->get_sort(), m_manager) << "\n";
-                  }
-                  else {
-                      enode * n = curr.get_enode();
-                      SASSERT(n->get_root() == n);
-                      sort * s = m_manager.get_sort(n->get_owner());
-                      tout << "#" << n->get_owner_id() << " " << mk_pp(s, m_manager);
-                      model_value_proc * proc = 0;
-                      root2proc.find(n, proc);
-                      SASSERT(proc);
-                      tout << " is_fresh: " << proc->is_fresh() << "\n";
-                  }
+                //ASHU: changed printing
+                source const & curr = *it;
+                if ( curr.is_fresh_value() ) {
+                  tout << "fresh!" << curr.get_value()->get_idx() << " "
+                       << mk_pp(curr.get_value()->get_sort(), m_manager)
+                       << "\n";
+                } else {
+                  enode * n = curr.get_enode();
+                  SASSERT(n->get_root() == n);
+                  sort * s = m_manager.get_sort(n->get_owner());
+                  tout << "#" << n->get_owner_id() << ":= ";
+                  ast_ll_bounded_pp( tout, m_manager, n->get_owner(), 1);
+                  tout << ":" << mk_pp( s, m_manager );
+                  model_value_proc * proc = 0;
+                  root2proc.find(n, proc);
+                  SASSERT(proc);
+                  tout << " is_fresh: " << proc->is_fresh() << "\n";
+                }
               });
         svector<source>::const_iterator it  = sources.begin();
         svector<source>::const_iterator end = sources.end();
@@ -533,5 +544,5 @@ namespace smt {
         register_macros();
         return m_model;
     }
-    
+
 };
diff --git a/src/smt/smt_model_generator.h b/src/smt/smt_model_generator.h
index 7466b8877..105dc90b5 100644
--- a/src/smt/smt_model_generator.h
+++ b/src/smt/smt_model_generator.h
@@ -32,6 +32,10 @@ Revision History:
 #include "smt/smt_types.h"
 #include "util/obj_hashtable.h"
 #include "util/map.h"
+//BEGIN: ASHU
+#include "model/func_interp.h"
+#include "ast/ast_ll_pp.h"
+//END: ASHU
 
 class value_factory;
 class proto_model;
@@ -184,6 +188,13 @@ namespace smt {
         proto_model *                 m_model;
         obj_hashtable<func_decl>      m_hidden_ufs;
 
+      //BEGIN: ASHU
+      func_decl_ref_vector f1s;
+      func_decl_ref_vector f2s;
+      expr_ref_vector added_constr;
+      expr_ref_vector assigned_atoms;
+      //END: ASHU
+
         void init_model();
         void mk_bool_model();
         void mk_value_procs(obj_map<enode, model_value_proc *> & root2proc, ptr_vector<enode> & roots,  ptr_vector<model_value_proc> & procs);
@@ -228,6 +239,11 @@ namespace smt {
                 m_manager.inc_ref(f); 
             }
         }
+      //BEGIN: ASHU
+      void collect_asserted_linear_constr( ast_ref_vector& );
+      void collect_asserted_po_constr( ast_ref_vector& );
+      app* search_matching_term( func_decl*, func_entry const*);
+      //END: ASHU
     };
 };
 
diff --git a/src/smt/smt_setup.cpp b/src/smt/smt_setup.cpp
index 55ea55663..15b092806 100644
--- a/src/smt/smt_setup.cpp
+++ b/src/smt/smt_setup.cpp
@@ -35,6 +35,7 @@ Revision History:
 #include "smt/theory_pb.h"
 #include "smt/theory_fpa.h"
 #include "smt/theory_str.h"
+#include "smt/theory_special_relations.h" // ASHU : added
 
 namespace smt {
 
@@ -889,6 +890,12 @@ namespace smt {
     void setup::setup_seq() {
         m_context.register_plugin(alloc(smt::theory_seq, m_manager));
     }
+  //BEGIN: ASHU
+    void setup::setup_special_relations() {
+        m_context.register_plugin(alloc(theory_special_relations, m_manager));
+        m_params.m_relevancy_lvl = 0;
+    }
+  //END: ASHU
 
     void setup::setup_unknown() {
         static_features st(m_manager);
@@ -904,6 +911,7 @@ namespace smt {
         setup_seq_str(st);
         setup_card();
         setup_fpa();
+        setup_special_relations();
     }
 
     void setup::setup_unknown(static_features & st) {
diff --git a/src/smt/smt_setup.h b/src/smt/smt_setup.h
index 924c2caec..49f361cdb 100644
--- a/src/smt/smt_setup.h
+++ b/src/smt/smt_setup.h
@@ -104,6 +104,7 @@ namespace smt {
         void setup_r_arith();
         void setup_fpa();
         void setup_str();
+        void setup_special_relations();
 
     public:
         setup(context & c, smt_params & params);
diff --git a/src/smt/smt_solver.cpp b/src/smt/smt_solver.cpp
index d282a59da..a1398823d 100644
--- a/src/smt/smt_solver.cpp
+++ b/src/smt/smt_solver.cpp
@@ -27,6 +27,11 @@ Notes:
 #include "ast/func_decl_dependencies.h"
 #include "util/dec_ref_util.h"
 
+//BEGIN: ASHU
+extern ast_ref_vector* asserted_linear_constr_vector;
+extern ast_ref_vector* asserted_po_constr_vector;
+//END: ASHU
+
 namespace smt {
 
     class smt_solver : public solver_na2as {
@@ -173,7 +178,14 @@ namespace smt {
 
         lbool check_sat_core(unsigned num_assumptions, expr * const * assumptions) override {
             TRACE("solver_na2as", tout << "smt_solver::check_sat_core: " << num_assumptions << "\n";);
-            return m_context.check(num_assumptions, assumptions);
+            //BEGIN: ASHU
+            lbool res = m_context.check(num_assumptions, assumptions);
+            asserted_linear_constr_vector->reset();
+            m_context.collect_asserted_linear_constr(*asserted_linear_constr_vector);
+            asserted_po_constr_vector->reset();
+            m_context.collect_asserted_po_constr(*asserted_po_constr_vector);
+            return res;
+            //END: ASHU
         }
 
         struct scoped_minimize_core {
diff --git a/src/smt/tactic/smt_tactic.cpp b/src/smt/tactic/smt_tactic.cpp
index cc0f0f207..ca2445d82 100644
--- a/src/smt/tactic/smt_tactic.cpp
+++ b/src/smt/tactic/smt_tactic.cpp
@@ -32,6 +32,11 @@ Notes:
 #include "solver/parallel_tactic.h"
 #include "solver/parallel_params.hpp"
 
+//BEGIN: ASHU
+extern ast_ref_vector* asserted_linear_constr_vector;
+extern ast_ref_vector* asserted_po_constr_vector;
+//END: ASHU
+
 typedef obj_map<expr, expr *> expr2expr_map;
 
 
@@ -139,6 +144,16 @@ public:
         ~scoped_init_ctx() {
             smt::kernel * d = m_owner.m_ctx;
             m_owner.m_ctx = nullptr;
+            //BEGIN: ASHU
+            if( asserted_linear_constr_vector ) {
+              asserted_linear_constr_vector->reset();
+              d->collect_asserted_linear_constr(*asserted_linear_constr_vector);
+            }
+            if( asserted_po_constr_vector ) {
+              asserted_po_constr_vector->reset();
+              d->collect_asserted_po_constr(*asserted_po_constr_vector);
+            }
+            //END: ASHU
 
             if (d)
                 dealloc(d);
diff --git a/src/smt/theory_arith.h b/src/smt/theory_arith.h
index 992a87dab..e8a5a05c3 100644
--- a/src/smt/theory_arith.h
+++ b/src/smt/theory_arith.h
@@ -1105,6 +1105,21 @@ namespace smt {
     public:
         void collect_statistics(::statistics & st) const override;
         void display(std::ostream & out) const override;
+
+  //BEGIN: ASHU
+      void collect_asserted_atoms( vector< std::pair<bool_var,bool> >& atoms) const {
+        for (unsigned i = 0; i < m_asserted_qhead; i++) {
+            bound * b = m_asserted_bounds[i];
+            if (b->is_atom()) {
+              atom* a = static_cast<atom*>(b);
+              bool_var bv = a->get_bool_var();
+              bool istr = a->is_true();
+              atoms.push_back( std::make_pair(bv,istr) );
+            }
+        }
+    }
+  //END: ASHU
+
     protected:
         void display_row(std::ostream & out, unsigned r_id, bool compact = true) const;
         void display_row(std::ostream & out, row const & r, bool compact = true) const;
